src/cir.rs

```diff
@@ -0,0 +1,437 @@
+                            ),
+                        };
+                        cir.instructions.push(CirInstruction::AndIns(and_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 3;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+                LirLogicalOperators::Or => {
+                    let next_ins = &lir.instructions[index + 1];
+                    if is_operator(next_ins) {
+                        // next is operator
+                        operator_stack.push(LirLogicalOperators::Or);
+                        index += 1;
+                    } else {
+                        let next_next_ins = &lir.instructions[index + 2];
+                        let or_ins = OrIns {
+                            left: CirOperand::Predicate(next_ins.as_predicate().unwrap().clone()),
+                            right: CirOperand::Predicate(
+                                next_next_ins.as_predicate().unwrap().clone(),
+                            ),
+                        };
+                        cir.instructions.push(CirInstruction::OrIns(or_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 3;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+                LirLogicalOperators::Not => {
+                    let next_ins = &lir.instructions[index + 1];
+                    if is_operator(next_ins) {
+                        // push LIR operator to ops stack
+                        // next is operator
+                        operator_stack.push(LirLogicalOperators::Not);
+                        index += 1;
+                    } else {
+                        next_ins.as_predicate().unwrap(); //right
+                        let not_ins = NotIns {
+                            right: CirOperand::Predicate(next_ins.as_predicate().unwrap().clone()),
+                        };
+                        cir.instructions.push(CirInstruction::NotIns(not_ins));
+                        operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
+                        index += 2;
+
+                        reduce_translation_stack(
+                            &mut cir.instructions,
+                            &mut operator_stack,
+                            &mut operand_stack,
+                        );
+                    }
+                }
+            },
```

Comment:
I would like to suggest to move out operator check on `next_ins` to reduce duplicated code.
And may using `BinaryIns` instead of `AndIns`/`OrIns` for better readibility.

```suggestion
            LirInstruction::LogicalOperator(op) => {
                let next_ins = &lir.instructions[index + 1];
                if is_operator(next_ins) {
                    // next is operator
                    operator_stack.push(*op); // Need LirLogicalOperators to implement Copy
                    index += 1;
                } else {
                    let (cir_instructions, operator_stack, operand_stack) = match op {
                        LirLogicalOperators::And | LirLogicalOperators::Or => {
                            let next_next_ins = &lir.instructions[index + 2];
                            cir.instructions.push(match op {
                                LirLogicalOperators::And => CirInstruction::AndIns(AndIns {
                                    left: CirOperand::Predicate(
                                        next_ins.as_predicate().unwrap().clone(),
                                    ),
                                    right: CirOperand::Predicate(
                                        next_next_ins.as_predicate().unwrap().clone(),
                                    ),
                                }),
                                LirLogicalOperators::Or => CirInstruction::OrIns(OrIns {
                                    left: CirOperand::Predicate(
                                        next_ins.as_predicate().unwrap().clone(),
                                    ),
                                    right: CirOperand::Predicate(
                                        next_next_ins.as_predicate().unwrap().clone(),
                                    ),
                                }),
                                _ => unreachable!(),
                            });
                            operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
                            index += 3;

                            (
                                &mut cir.instructions,
                                &mut operator_stack,
                                &mut operand_stack,
                            )
                        }
                        LirLogicalOperators::Not => {
                            next_ins.as_predicate().unwrap(); //right
                            let not_ins = NotIns {
                                right: CirOperand::Predicate(
                                    next_ins.as_predicate().unwrap().clone(),
                                ),
                            };
                            cir.instructions.push(CirInstruction::NotIns(not_ins));
                            operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
                            index += 2;

                            (
                                &mut cir.instructions,
                                &mut operator_stack,
                                &mut operand_stack,
                            )
                        }
                    };
                    reduce_translation_stack(cir_instructions, operator_stack, operand_stack);
                }
            },
```

Reply:
fixed, good point, much better. thanks.

Reply:
Here is a better version if the latest commit does not include this. And this avoids panic when lir is empty.

```Rust
fn cir_translate_helper(lir: &LirProgram, cir: &mut CirProgram) {
    let mut operand_stack: Vec<CirOperand> = Vec::new();
    let mut operator_stack: Vec<LirLogicalOperators> = Vec::new();
    let mut index = 0;
    loop {
        if index >= lir.instructions.len() {
            // end of LirProgram
            break;
        }

        match &lir.instructions[index] {
            LirInstruction::LogicalOperator(op) => {
                let next_ins = &lir.instructions[index + 1];
                if is_operator(next_ins) {
                    // next is operator
                    operator_stack.push(*op); // Need LirLogicalOperators to implement Copy
                    index += 1;
                    continue;
                } else {
                    match op {
                        // Binary experssion
                        LirLogicalOperators::And | LirLogicalOperators::Or => {
                            let next_next_ins = &lir.instructions[index + 2];
                            cir.instructions.push(match op {
                                LirLogicalOperators::And => CirInstruction::AndIns(AndIns {
                                    left: CirOperand::Predicate(next_ins.as_predicate().clone()),
                                    right: CirOperand::Predicate(
                                        next_next_ins.as_predicate().clone(),
                                    ),
                                }),
                                LirLogicalOperators::Or => CirInstruction::OrIns(OrIns {
                                    left: CirOperand::Predicate(next_ins.as_predicate().clone()),
                                    right: CirOperand::Predicate(
                                        next_next_ins.as_predicate().clone(),
                                    ),
                                }),
                                _ => unreachable!(),
                            });
                            operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
                            index += 3;
                        }
                        // Unary expression
                        LirLogicalOperators::Not => {
                            next_ins.as_predicate(); //right
                            let not_ins = NotIns {
                                right: CirOperand::Predicate(next_ins.as_predicate().clone()),
                            };
                            cir.instructions.push(CirInstruction::NotIns(not_ins));
                            operand_stack.push(CirOperand::Index(cir.instructions.len() - 1));
                            index += 2;
                        }
                    };
                }
            }
            LirInstruction::Predicate(p) => {
                operand_stack.push(CirOperand::Predicate(p.clone()));
                index += 1;
            }
        }

        // End of a logical block
        reduce_translation_stack(
            &mut cir.instructions,
            &mut operator_stack,
            &mut operand_stack,
        );
    }
    debug_assert_eq!(operator_stack.len(), 0);
}
```

Reply:
fixed.